core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -0,0 +1,297 @@
+      "failed to get expected partition state upon broker shutdown")
+  }
+
+  @Test
+  def testAutoPreferredReplicaLeaderElection(): Unit = {
+    servers = makeServers(2, autoLeaderRebalanceEnable = true)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch + 1,
+      "failed to get expected partition state upon broker shutdown")
+    servers(1).startup()
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 1, LeaderAndIsr.initialLeaderEpoch + 2,
+      "failed to get expected partition state upon broker startup")
+  }
+
+  @Test
+  def testLeaderAndIsrWhenEntireIsrOfflineAndUncleanLeaderElectionDisabled(): Unit = {
+    servers = makeServers(2)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 1, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    TestUtils.waitUntilTrue(() => {
+      val leaderIsrAndControllerEpochMap = zkUtils.getPartitionLeaderAndIsrForTopics(null, Set(tp))
+      leaderIsrAndControllerEpochMap.contains(tp) &&
+        isExpectedPartitionState(leaderIsrAndControllerEpochMap(tp), KafkaController.InitialControllerEpoch, LeaderAndIsr.NoLeader, LeaderAndIsr.initialLeaderEpoch + 1) &&
+        leaderIsrAndControllerEpochMap(tp).leaderAndIsr.isr == List(1)
+    }, "failed to get expected partition state after entire isr went offline")
+  }
+
+  @Test
+  def testLeaderAndIsrWhenEntireIsrOfflineAndUncleanLeaderElectionEnabled(): Unit = {
+    servers = makeServers(2, uncleanLeaderElectionEnable = true)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 1, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    TestUtils.waitUntilTrue(() => {
+      val leaderIsrAndControllerEpochMap = zkUtils.getPartitionLeaderAndIsrForTopics(null, Set(tp))
+      leaderIsrAndControllerEpochMap.contains(tp) &&
+        isExpectedPartitionState(leaderIsrAndControllerEpochMap(tp), KafkaController.InitialControllerEpoch, LeaderAndIsr.NoLeader, LeaderAndIsr.initialLeaderEpoch + 1) &&
+        leaderIsrAndControllerEpochMap(tp).leaderAndIsr.isr == List.empty
```

Comment:
Hmm, it's probably better to leave the last replica in isr in ZK, independent of whether unclean leader election is enabled or not. This way, if people re-disabled unclean leader election, we can still try to elect the leader from the last in-sync replica. So, we probably want to fix the controller logic to do that.